File: DynamicProgramming/maxProfit2/main.cpp
#include <vector>

using namespace std;

class Solution {
 public:
  int maxProfit(vector<int>& prices) {
    vector<int> maxProfitTable = vector<int>(prices.size(), 0);

    for (int i = 1; i < prices.size(); i++) {
      int sellToday = prices[i] - prices[i - 1] + maxProfitTable[i - 1];
      int doNotSellToday = maxProfitTable[i - 1];
      maxProfitTable[i] = max(sellToday, doNotSellToday);
    }

    return maxProfitTable[prices.size() - 1];
  }
};

class Solution {
 public:
  int maxProfit(vector<int>& prices) {
    int maxProfit = 0;

    for (int i = 1; i < prices.size(); i++) {
      int sellToday = prices[i] - prices[i - 1] + maxProfit;
      int doNotSellToday = maxProfit;
      maxProfit = max(sellToday, doNotSellToday);
    }

    return maxProfit;
  }
};

File: DynamicProgramming/maxProfit2/go/main.go
package main

func MaxProfit(prices []int) int {
	maxProfitTable := make([]int, len(prices))

	// start from the second day
	// because we know the first day's profit is always 0
	for i := 1; i < len(prices); i++ {
		sellToday := prices[i] - prices[i-1] + maxProfitTable[i-1]
		doNotSellToday := maxProfitTable[i-1]
		maxProfitTable[i] = max(sellToday, doNotSellToday)
	}

	return maxProfitTable[len(prices)-1]
}

func MaxProfit1(prices []int) int {
	maxProfit := 0

	// start from the second day
	// because we know the first day's profit is always 0
	for i := 1; i < len(prices); i++ {
		sellToday := prices[i] - prices[i-1] + maxProfit
		doNotSellToday := maxProfit
		maxProfit = max(sellToday, doNotSellToday)
	}

	return maxProfit
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
